build_tools/cmake/build_runtime_emscripten.sh

```diff
@@ -0,0 +1,82 @@
+#!/bin/bash
+# Copyright 2021 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+# Cross-compile IREE's runtime through Emscripten to WebAssembly with CMake.
+# Designed for CI, but can be run manually. This uses previously cached build
+# results and does not clear build directories.
+#
+# Host binaries (e.g. compiler tools) will be built and installed in build-host/
+# Emscripten binaries (e.g. .wasm and .js files) will be built in
+# build-emscripten/.
+
+set -x
+set -e
+
+if ! command -v emcmake &> /dev/null
+then
+    echo "'emcmake' not found, setup environment according to https://emscripten.org/docs/getting_started/downloads.html"
+    exit
+fi
+
+ROOT_DIR=$(git rev-parse --show-toplevel)
+
+CMAKE_BIN=${CMAKE_BIN:-$(which cmake)}
+
+"${CMAKE_BIN?}" --version
+ninja --version
+
+cd ${ROOT_DIR?}
+
+# --------------------------------------------------------------------------- #
+# Build for the host.
+
+if [ -d "build-host" ]
+then
+  echo "build-host directory already exists. Will use cached results there."
+else
+  echo "build-host directory does not already exist. Creating a new one."
+  mkdir build-host
+fi
+cd build-host
+
+# Configure, build, install.
+# Just build the host components that we need for an Emscripten runtime build.
+"${CMAKE_BIN?}" -G Ninja .. \
+  -DCMAKE_INSTALL_PREFIX=./install \
+  -DIREE_BUILD_COMPILER=ON \
+  -DIREE_TARGET_BACKENDS_TO_BUILD=VMVX\;WASM-LLVM-AOT \
+  -DIREE_HAL_DRIVERS_TO_BUILD= \
+  -DIREE_BUILD_TESTS=OFF \
+  -DIREE_BUILD_SAMPLES=OFF
+"${CMAKE_BIN?}" --build . --target install
```

Comment:
@GMNGeoffrey this uses different settings compared to https://github.com/google/iree/blob/main/build_tools/cmake/build_riscv.sh and https://github.com/google/iree/blob/main/build_tools/cmake/build_android.sh 

I also see errors like this when switching between scripts and docker runs:
```
CMake Error: The current CMakeCache.txt directory /usr/local/google/home/scotttodd/code/iree/build-host/CMakeCache.txt is different than the directory /usr/local/google/home/scotttodd/tmp/docker/iree/build-host where CMakeCache.txt was created. This may result in binaries being created in the wrong place. If you are not sure, reedit the CMakeCache.txt
CMake Error: The source "/usr/local/google/home/scotttodd/code/iree/CMakeLists.txt" does not match the source "/usr/local/google/home/scotttodd/tmp/docker/iree/CMakeLists.txt" used to generate cache.  Re-run cmake with a different source directory.
```

any ideas on how to keep this organized as the number of configurations scales?

Reply:
Maybe we could pick a set of common options and pull those into a `build_host_install.sh` script?

Reply:
I think we'd want to pull it out at the point where we had a separate build for the host that then provided artifacts shared by other builds. Otherwise, it seems good to, for instance, not build the wasm backend for the android tests. Although maybe the win from not forgetting to turn off all the iree hal drivers is worth the overhead :-)

Reply:
Oo yeah, wiring it up as a multi-stage build within BuildKite would be great. That's not quite what I had in mind for making these scripts generally usable by developers, but it makes more sense in the long run to spend our energy there.

The disabling of compiler targets and runtime drivers are mostly small optimizations and ways to enforce clean boundaries between components. The entire runtime build is fast (~1 minute), and the webassembly compiler target is essentially the same as the llvm cpu compiler target, except for some different codegen deps: https://github.com/google/iree/blob/5057f885ae45aef38f20d926eabced6ab80397c3/iree/compiler/Dialect/HAL/Target/LLVM/CMakeLists.txt#L40-L43
and calls to `INIT_LLVM_TARGET`: https://github.com/google/iree/blob/5057f885ae45aef38f20d926eabced6ab80397c3/iree/compiler/Dialect/HAL/Target/LLVM/LLVMAOTTarget.cpp#L523-L530
(which are all unconditional right now, so we don't actually save time by trying to disable at that granularity).

Reply:
Cool, I like this direction. I'll switch this `build_configurations` pipeline to use https://buildkite.com/docs/pipelines/dependencies#defining-explicit-dependencies and a setup like https://github.com/google/iree/blob/5057f885ae45aef38f20d926eabced6ab80397c3/build_tools/buildkite/cmake/android/arm64-v8a/pipeline.yml#L11-L23

for passing artifacts between stages. I'll define a common `build_host_install.sh` script + BuildKite step that the `build_runtime_emscripten.sh` script (and possibly `build_riscv.sh` and `build_android` scripts) will run after.

Once _this_ pipeline has the multistage host / runtime configuration build working, we can update our other pipelines in similar ways.

Reply:
Okay, that works: https://buildkite.com/iree/scotttodd-test-pipeline/builds/48
Future cleanup can chain other builds off of that within this pipeline or rework our other pipelines.